sdk/core/azure-core/azure/core/rest/_http_response_impl.py

```diff
@@ -0,0 +1,320 @@
+        self._status_code = kwargs.pop("status_code")  # type: int
+        self._reason = kwargs.pop("reason")  # type: str
+        self._content_type = kwargs.pop("content_type")  # type: str
+        self._headers = kwargs.pop("headers")  # type: MutableMapping[str, str]
+        self._stream_download_generator = kwargs.pop("stream_download_generator")  # type: Callable
+        self._is_closed = False
+        self._is_stream_consumed = False
+        self._json = None  # this is filled in ContentDecodePolicy, when we deserialize
+        self._content = None  # type: Optional[bytes]
+        self._text = None  # type: Optional[str]
+
+    @property
+    def request(self):
+        # type: (...) -> _HttpRequest
+        """The request that resulted in this response.
+
+        :rtype: ~azure.core.rest.HttpRequest
+        """
+        return self._request
+
+    @property
+    def url(self):
+        # type: (...) -> str
+        """The URL that resulted in this response.
+
+        :rtype: str
+        """
+        return self.request.url
+
+    @property
+    def is_closed(self):
+        # type: (...) -> bool
+        """Whether the network connection has been closed yet.
+
+        :rtype: bool
+        """
+        return self._is_closed
+
+    @property
+    def is_stream_consumed(self):
+        # type: (...) -> bool
+        """Whether the stream has been fully consumed"""
+        return self._is_stream_consumed
+
+    @property
+    def status_code(self):
+        # type: (...) -> int
+        """The status code of this response.
+
+        :rtype: int
+        """
+        return self._status_code
+
+    @property
+    def headers(self):
+        # type: (...) -> MutableMapping[str, str]
+        """The response headers.
+
+        :rtype: MutableMapping[str, str]
+        """
+        return self._headers
+
+    @property
+    def content_type(self):
+        # type: (...) -> str
+        """The content type of the response.
+
+        :rtype: str
+        """
+        return self._content_type
+
+    @property
+    def reason(self):
+        # type: (...) -> str
+        """The reason phrase for this response.
+
+        :rtype: str
+        """
+        return self._reason
+
+    @property
+    def encoding(self):
+        # type: (...) -> Optional[str]
+        """Returns the response encoding.
+
+        :return: The response encoding. We either return the encoding set by the user,
+         or try extracting the encoding from the response's content type. If all fails,
+         we return `None`.
+        :rtype: optional[str]
+        """
+        try:
+            return self._encoding
+        except AttributeError:
+            self._encoding = get_charset_encoding(self)  # type: Optional[str]
+            return self._encoding
+
+    @encoding.setter
+    def encoding(self, value):
+        # type: (str) -> None
+        """Sets the response encoding"""
+        self._encoding = value
```

Comment:
Maybe 
```py
if self._encoding != value:
    self._encoding = value
    self._text = None  # clear text cache
```

Reply:
if setting encoding resets self._text, do we want to reset self._json as well?

Reply:
that's a good point, let me do that!